server/src/main/java/org/candlepin/model/Product.java

```diff
@@ -939,7 +1077,174 @@ public Product setLocked(boolean locked) {
+            // collections explicitly state that they break the contract on .equals. As such, we
+            // have to step through each collection and do a manual comparison. Ugh.
+            equals = equals &&
+                Util.collectionsAreEqual(this.attributes, that.attributes) &&
+                Util.collectionsAreEqual(this.dependentProductIds, that.dependentProductIds) &&
+                Util.collectionsAreEqual(this.productContent, that.productContent);
+        }
+
+        return equals;
+    }
+
+    @Override
+    public int hashCode() {
+        // This must always be a subset of equals
+        HashCodeBuilder builder = new HashCodeBuilder(37, 7)
+            .append(this.id)
+            .append(this.name)
+            .append(this.multiplier)
+            .append(this.locked);
+
+        // Impl note:
+        // Because we handle the collections specially in .equals, we have to do the same special
+        // treatment here to ensure our output doesn't give us wonky results when compared to the
+        // output of .equals
+        for (ProductAttribute attrib : this.attributes) {
+            builder.append(attrib);
+        }
+
+        try {
+            // Impl note:
+            // Stepping through the collections here is as painful as it looks, but Hibernate, once
+            // again, doesn't implement .hashCode reliably on the proxy collections. So, we have to
+            // manually step through these and add the elements to ensure the hash code is
+            // generated properly.
+            if (this.dependentProductIds.size() > 0) {
+                for (String pid : this.dependentProductIds) {
+                    builder.append(pid);
+                }
+            }
+
+            if (this.productContent.size() > 0) {
+                for (ProductContent pc : this.productContent) {
+                    builder.append(pc);
+                }
+            }
+        }
+        catch (LazyInitializationException e) {
+            // One of the above collections (likely the first) has not been initialized and we're
+            // not able to fetch them. We still need a hashCode, and the caller is likely to run
+            // into this exception in the very near future, but we still need to generate
+            // something here. We'll treat it as if they were empty (and not added).
+
+            // This typically only occurs when we're initially pulling the entity down from a normal
+            // lookup. Hibernate stores the entity in a hashmap, which triggers a call to this
+            // method before Hibernate has fully hydrated it and before it's assigned it to an
+            // entity manager or session. As such, as soon as we try to use one of the above
+            // collections, things explode.
+
+            // Note that this only applies to lazy collections, so the product attributes are safe
+            // to check for now.
```

Comment:
That check's been in master for a while now -- probably just an indentation change. The only time the exception is thrown is during hydration before Hibernate gets it back. So, I can add some debug output here, but it'll be mostly noise.